hibernate-core/src/main/java/org/hibernate/bytecode/enhance/spi/interceptor/EnhancementHelper.java

```diff
@@ -51,15 +58,38 @@ public static boolean includeInBaseFetchGroup(
 		if ( value instanceof ToOne ) {
 			final ToOne toOne = (ToOne) value;
 			if ( toOne.isLazy() ) {
-				if ( toOne.isUnwrapProxy() ) {
-					if ( toOne instanceof OneToOne ) {
-						return false;
+				// include it in the base fetch group so long as the config allows
+				// using the FK to create an "enhancement proxy"
+
+				// however a few special cases to handle...
+
+				if ( ! toOne.isReferenceToPrimaryKey() ) {
+					// we do not have a reference to the associated primary-key
+					return false;
+				}
+
+				if ( inheritanceChecker.hasSubclasses( toOne.getReferencedEntityName() ) ) {
+					// the associated type has subclasses - we cannot use the enhanced proxy and will generate a HibernateProxy
+					if ( ! toOne.isUnwrapProxyImplicit() ) {
+						// NO_PROXY was explicitly requested
+						BytecodeLogger.LOGGER.debugf(
+								"`%s#%s` was mapped with LAZY with explicit NO_PROXY but the associated entity (`%s`) has subclasses",
+								bootMapping.getPersistentClass().getEntityName(),
+								bootMapping.getName(),
+								toOne.getReferencedEntityName()
+						);
 					}
-					// include it in the base fetch group so long as the config allows
-					// using the FK to create an "enhancement proxy"
+					// however, select the fk to create the proxy
+					return true;
+				}
+
+				if ( toOne.isUnwrapProxy() && ! toOne.isUnwrapProxyImplicit() ) {
```

Comment:
I'm unsure about this `if` statement. I'm going to do more checking on this Monday.

Reply:
This check is actually the exact improvement.  Previously, if `@LazyToOne` was not specified (implicit) then unwrap-proxy would be set to true.  However, if `@LazyToOne` is implicit we want to treat that differently.

Reply:
OK, I understand this case (explicitly has NO_PROXY) now, so this looks OK.